omp/components/prefix_sum.cpp

```diff
@@ -33,22 +33,67 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/components/prefix_sum.hpp"
 
 
+#include <algorithm>
+#include <vector>
+
+
+#include <omp.h>
+
+
 namespace gko {
 namespace kernels {
 namespace omp {
 namespace components {
 
 
+/*
+ * The last entry of the input array is never used, but is replaced.
+ */
 template <typename IndexType>
-void prefix_sum(std::shared_ptr<const OmpExecutor> exec, IndexType *counts,
-                size_type num_entries)
+void prefix_sum(std::shared_ptr<const OmpExecutor> exec,
+                IndexType *const counts, const size_type num_entries)
 {
-    IndexType partial_sum{};
-    for (size_type i = 0; i < num_entries; ++i) {
-        auto nnz = counts[i];
-        counts[i] = partial_sum;
-        partial_sum += nnz;
+    const auto nentries = static_cast<IndexType>(num_entries);
+    // the operation only makes sense for arrays of size at least 2
+    GKO_ENSURE_COMPATIBLE_BOUNDS(2, nentries);
+
+    const int nthreads = omp_get_max_threads();
+    std::vector<IndexType> proc_sums(nthreads, 0);
+    const IndexType def_num_witems = (num_entries - 1) / nthreads + 1;
```

Comment:
Random thought: Can nthreads ever be 0? Does the runtime make any guarantees for mis-configured OMP environments?

Reply:
I don't think so. All I could find in the spec (4.5) is
> The return value of the `omp_get_max_threads` routine can be used to dynamically allocate sufficient storage for all threads in the team formed at the subsequent active parallel region.

If the OpenMP implementation is correct, there's no way for `nthreads` to be zero or less. Should we worry about improperly configured OpenMP? I guess I could add a check and set it to 1 by default. I could add that to omp/components or something.